flink-datastream/src/main/java/org/apache/flink/datastream/impl/operators/ProcessOperator.java

```diff
@@ -98,6 +100,21 @@ protected Object currentKey() {
         throw new UnsupportedOperationException("The key is only defined for keyed operator");
     }
 
+    protected BiConsumer<Runnable, Object> getProcessorWithKey() {
+        if (isAsyncStateProcessingEnabled()) {
+            return (r, k) -> asyncProcessWithKey(k, r::run);
```

Comment:
Does this make the `executeInKeyContext` an asynchronous call?

Reply:
Yes. The input `runnable` will act just like a `processElement`. The CPU part would still run in task thread and it allows user to perform async state requests in their `runnable`, after which the provided state callback will run back in task thread.